src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityWorkbench.java

```diff
@@ -190,34 +194,233 @@ public void clearMachineInventory(@NotNull List<@NotNull ItemStack> itemBuffer)
-        builder.label(5, 5, getMetaFullName());
+    public IWidget createToolInventory(PanelSyncManager syncManager) {
+        var toolSlots = new SlotGroup("tool_slots", 9, -120, true);
+        syncManager.registerSlotGroup(toolSlots);
+
+        return SlotGroupWidget.builder()
+                .row("XXXXXXXXX")
+                .key('X', i -> new ItemSlot()
+                        .background(GTGuiTextures.SLOT, GTGuiTextures.TOOL_SLOT_OVERLAY)
+                        .slot(SyncHandlers.itemSlot(this.toolInventory, i)
+                                .slotGroup(toolSlots)))
+                .build().marginTop(2);
+    }
+
+    public IWidget createInternalInventory(PanelSyncManager syncManager) {
+        var inventory = new SlotGroup("internal_slots", 9, -100, true);
+        syncManager.registerSlotGroup(inventory);
+
+        return SlotGroupWidget.builder()
+                .row("XXXXXXXXX")
+                .row("XXXXXXXXX")
+                .key('X', i -> new ItemSlot()
+                        .slot(SyncHandlers.itemSlot(this.internalInventory, i)
+                                .slotGroup(inventory)))
+                .build().marginTop(2);
+    }
+
+    public IWidget createCraftingGrid() {
+        return SlotGroupWidget.builder()
+                .matrix("XXX",
+                        "XXX",
+                        "XXX")
+                .key('X', i -> CraftingInputSlot.create(this.recipeLogic, this.craftingGrid, i)
+                        .changeListener((newItem, onlyAmountChanged, client, init) -> {
+                            if (!init) {
+                                this.recipeLogic.updateCurrentRecipe();
+                            }
+                        })
+                        .background(GTGuiTextures.SLOT))
+                .build()
+                .child(new ButtonWidget<>()
+                        .margin(2)
+                        .size(8)
+                        .topRel(0f)
+                        .rightRel(0f, 0, 1f)
+                        .background(GTGuiTextures.BUTTON_CLEAR_GRID)
+                        .addTooltipLine(IKey.lang("gregtech.machine.workbench.clear_grid"))
+                        .disableHoverBackground()
+                        .onMousePressed(mouseButton -> {
+                            this.recipeLogic.clearCraftingGrid();
+                            this.recipeLogic.syncToServer(
+                                    CraftingRecipeLogic.UPDATE_MATRIX,
+                                    this.recipeLogic::writeMatrix);
+                            return true;
+                        }));
+    }
+
+    public IWidget createCraftingOutput(PosGuiData guiData, PanelSyncManager syncManager) {
+        var amountCrafted = new IntSyncValue(this::getItemsCrafted, this::setItemsCrafted);
+        syncManager.syncValue("amount_crafted", amountCrafted);
+        amountCrafted.updateCacheFromSource(true); // todo remove
```

Comment:
remove?

Reply:
this is to work around an issue with int sync values where they don't set their cache on construction, there's a mixin in #2672 but it's also fixed in mui2 rc3